### ISSUE TITLE
Problem trying to restart on pm-cpu/pm-gpu

### ISSUE THREAD

**Issue by ndkeen** [E3SM-Project/scream#1934]
I've had some issues trying to write/read restarts, but has not been clear what's happening as not always repeatable.
I'm not sure if this is the same thing, but it was a simple case of running 5 years, writing restart, stopping.  Then trying to CONTINUE_RUN for another 5 years.

```
 548: terminate called after throwing an instance of 'std::logic_error'
 548:   what():  /global/cfs/cdirs/e3sm/ndk/se28-sep8/components/scream/src/share/io/scream_io_utils.cpp:66: FAIL:
 548: found
 548: Error! Restart requested, but no restart file found in 'rpointer.atm'.
 548:    restart case name: f30cpu.F2010-SCREAMv1.ne30pg2_ne30pg2.se28-sep8.gnu.5y.n022a128x1cXi2.pk1.so480n2.scream
 548:    restart file type: model restart
 548:    rpointer content:
```

In fact, there does not appear to be a `rpointer.atm` file at all.

`/global/cfs/cdirs/e3sm/ndk/e3sm_scratch/pm-cpu/se28-sep8/f30cpu.F2010-SCREAMv1.ne30pg2_ne30pg2.se28-sep8.gnu.5y.n022a128x1cXi2.pk1.so480n2`

**Comment by AaronDonahue** [comment]
@bartgol , haven't we seen something like this before, with not writing the rpointer file?

**Comment by bartgol** [comment]
This is new to me. Did the initial simulation run to completion, and exited without any error? Just trying to rule out issues that could have prevented scorpio from actually writing the file...

**Comment by ndkeen** [comment]
Yes the original run did complete -- gzipped the log files, reported 38.9 sypd

But clearly, it didn't seem to write all the restart "stuff":

```
perlmutter% lr run/*.r.*
-rw-rw-r-- 1 ndk e3sm   6060120 Sep  9 05:08 run/f30cpu.F2010-SCREAMv1.ne30pg2_ne30pg2.se28-sep8.gnu.5y.n022a128x1cXi2.pk1.so480n2.cice.r.0006-01-01-00000.nc
-rw-rw-r-- 1 ndk e3sm 401303276 Sep  9 05:08 run/f30cpu.F2010-SCREAMv1.ne30pg2_ne30pg2.se28-sep8.gnu.5y.n022a128x1cXi2.pk1.so480n2.elm.r.0006-01-01-00000.nc
-rw-rw-r-- 1 ndk e3sm  39546000 Sep  9 05:08 run/f30cpu.F2010-SCREAMv1.ne30pg2_ne30pg2.se28-sep8.gnu.5y.n022a128x1cXi2.pk1.so480n2.cpl.r.0006-01-01-00000.nc
perlmutter% lr run/rpo*
-rw-rw-r-- 1 ndk e3sm 257 Sep  9 05:08 run/rpointer.ice
-rw-rw-r-- 1 ndk e3sm 257 Sep  9 05:08 run/rpointer.lnd
-rw-rw-r-- 1 ndk e3sm 514 Sep  9 05:08 run/rpointer.ocn
-rw-rw-r-- 1 ndk e3sm 257 Sep  9 05:08 run/rpointer.drv
```

**Comment by bartgol** [comment]
Ok, a couple of question to guide debugging:

- you said it's not always reproducible, right?
- were there any changes (either XML or atmchange) to the defaults?

**Comment by ndkeen** [comment]
Correct -- with other cases (including slightly diff repos), I've had mixed results -- writing a single restart and reading it works most of the time -- but at least twice, trying to write 2 restarts in a simulation had trouble (perhaps not same issue). 

I've been running various cases with pretty much same setup -- only making minor changes (for example PE layout, or scream packsize, etc)

**Comment by ndkeen** [comment]
Had at least 2 more cases (this time on pm-gpu) have this issue -- where the restart seems to complete (?), but there is no rpointer.atm file.

I have also tried several smaller tests that have worked -- ie on both pm-cpu/pm-gpu, run ne30 for a month, restart for a month -- repeat several times.

And here is example of a year long case on cori-knl that seemed to have worked as well.
`/global/cscratch1/sd/ndk/e3sm_scratch/cori-knl/se29-sep11/f30cpu.F2010-SCREAMv1.ne30pg2_ne30pg2.se29-sep11.intel.n043a64x2c4.pk8.w500`

So it's not something as simple as "ok-with-months-but-not-years".

**Comment by ndkeen** [comment]
Another failure mode. I tried ne256 on pm-gpu with master of Sep 19th. I tried to run for 1 day and write a restart file.


```
  0: Create Pool
  0: Create Pool
  0: PIO: FATAL ERROR: Aborting... FATAL ERROR: NetCDF: NC_UNLIMITED size already in use (file = ./f256.F2010-SCREAMv1-noAero.ne256pg2_r0125_oRRS18to6v3.se30-sep19.gnugpu.1d.n048a4x16ci2.IC914.so144n1.wr.elm.r.0001-01-02-00000.nc) (/global/cfs/cdirs/e3sm/ndk/se30-sep19/externals/scorpio/src/clib/pio_nc.c: 2770)
  0: Obtained 10 stack frames.
  0: /global/cfs/cdirs/e3sm/ndk/e3sm_scratch/pm-gpu/se30-sep19/f256.F2010-SCREAMv1-noAero.ne256pg2_r0125_oRRS18to6v3.se30-sep19.gnugpu.1d.n048a4x16ci2.IC914.so144n1.wr/bld/e3sm.exe() [0xf47815]
  0: /global/cfs/cdirs/e3sm/ndk/e3sm_scratch/pm-gpu/se30-sep19/f256.F2010-SCREAMv1-noAero.ne256pg2_r0125_oRRS18to6v3.se30-sep19.gnugpu.1d.n048a4x16ci2.IC914.so144n1.wr/bld/e3sm.exe() [0xf4795e]
  0: /global/cfs/cdirs/e3sm/ndk/e3sm_scratch/pm-gpu/se30-sep19/f256.F2010-SCREAMv1-noAero.ne256pg2_r0125_oRRS18to6v3.se30-sep19.gnugpu.1d.n048a4x16ci2.IC914.so144n1.wr/bld/e3sm.exe() [0xf47d7d]
  0: /global/cfs/cdirs/e3sm/ndk/e3sm_scratch/pm-gpu/se30-sep19/f256.F2010-SCREAMv1-noAero.ne256pg2_r0125_oRRS18to6v3.se30-sep19.gnugpu.1d.n048a4x16ci2.IC914.so144n1.wr/bld/e3sm.exe() [0xf5daa5]

/global/cfs/cdirs/e3sm/ndk/e3sm_scratch/pm-gpu/se30-sep19/f256.F2010-SCREAMv1-noAero.ne256pg2_r0125_oRRS18to6v3.se30-sep19.gnugpu.1d.n048a4x16ci2.IC914.so144n1.wr
```

Note that I still need to turn off the writing of `elm.h0` files, or I will see a hang.  
```
hist_nhtfrq = -99999999
hist_mfilt = 1
hist_empty_htapes = .true.
```

**Comment by bartgol** [comment]
It would be nice if PIO could print the dim name, as well as the other dim name with ulimited size already registered.

**Comment by bartgol** [comment]
Is this resolution specific? Does it show also at ne4? Would be easy to debug. I guess I can check myself.

**Comment by ambrad** [comment]
I'm wondering if you need 64bit_data rather than 64bit_offset for ne256pg2.

**Comment by ndkeen** [comment]
Ahh thanks Andrew. Indeed I use `xmlchange PIO_NETCDF_FORMAT="64bit_data"` with ne1024 scripts, but did not have it for this ne256 attempt with restarts.  Trying again.

**Comment by ndkeen** [comment]
Well that was not the issue. I tried again with `64bit_data` and get the same error.

**Comment by ndkeen** [comment]
This could easily be 2 different issues.  With ne30, I was able to run 6 months with a restart every month -- have not tested a read, but the rpointer.atm is present. I also tried running multiple years and writing a restart every year which worked for the first 2 years and then failed with high wind speed (known issue at ne30). Having trouble reproducing the error at ne30.

**Comment by ndkeen** [comment]
With ne120, on pm-gpu, using 24 nodes, I tried to run 1 day and write a restart. It fails with the `FATAL ERROR: NetCDF: NC_UNLIMITED size already in use` error.  The same case will complete 1 day when I don't ask for restarts.

**Comment by ndkeen** [comment]
I tried to run 10years of ne30 and write a restart every year on pm-gpu. This time it failed writing the first restart, but it's not the same fail pattern as before.  It looks like it stopped trying to write `cpl.r` and wrote an empty `rpointer.drv` file. 
`/global/cfs/cdirs/e3sm/ndk/e3sm_scratch/pm-gpu/se31-sep21/f30.F2010-SCREAMv1.ne30pg2_ne30pg2.se31-sep21.gnugpu.10y.n008a4xX14444.so480n8.w500.wr`

**Comment by ndkeen** [comment]
@AaronDonahue reports that on summit a ne120 case ran for 1 month writing restarts every day. That included a `elm.h0` file. He also did a similar experiment with ne256 that timed out, but appears to be writing restarts ok.

**Comment by ndkeen** [comment]
I tried a ne120 test case on a machine very similar to pm-cpu.  This used scream master of Oct 10th _and_ the latest scorpio master (of Oct 11th -- ie I did `git pull origin master` inside of externals/scorpio).

It works without restart attempt, but with it, I see:

```
 256: PIO: FATAL ERROR: Aborting... FATAL ERROR: NetCDF: NC_UNLIMITED size already in use (file = ./f120.F2010-SCREAMv1.ne120_r0125_oRRS18to6v3.se36-oct10.gnu.5d.n043a128x1c8.pk8.noew.wr.elm.r.0001-01-06-00000.nc) (/global/cfs/cdirs/e3sm/ndk/se36-oct10/externals/scorpio/src/clib/pio_nc.c: 2863)
 256: Obtained 10 stack frames.
 256: /ascratch/sd/n/ndk/e3sm_scratch/alvarez/se36-oct10/f120.F2010-SCREAMv1.ne120_r0125_oRRS18to6v3.se36-oct10.gnu.5d.n043a128x1c8.pk8.noew.wr/bld/e3sm.exe() [0xf08335]
 256: /ascratch/sd/n/ndk/e3sm_scratch/alvarez/se36-oct10/f120.F2010-SCREAMv1.ne120_r0125_oRRS18to6v3.se36-oct10.gnu.5d.n043a128x1c8.pk8.noew.wr/bld/e3sm.exe() [0xf0847e]
 256: /ascratch/sd/n/ndk/e3sm_scratch/alvarez/se36-oct10/f120.F2010-SCREAMv1.ne120_r0125_oRRS18to6v3.se36-oct10.gnu.5d.n043a128x1c8.pk8.noew.wr/bld/e3sm.exe() [0xf0889d]
 256: /ascratch/sd/n/ndk/e3sm_scratch/alvarez/se36-oct10/f120.F2010-SCREAMv1.ne120_r0125_oRRS18to6v3.se36-oct10.gnu.5d.n043a128x1c8.pk8.noew.wr/bld/e3sm.exe() [0xf1e655]
```

**Comment by ndkeen** [comment]
I don't think we have this specific issue anymore (though I'm not sure what, if anything, changed). And I'm not sure if we want to submit some longer ne30 runs just to verify. With charging starting on PM, maybe we can find another reason to make some longer runs and then I can verify no issue with restarts as a secondary benefit.

**Comment by jayeshkrishna** [comment]
I remember @whannah1 mentioning a similar issue that he saw when trying to disable elm output (when the ELM tapes were set to 0, the code was incorrectly setting the dimension length for a dimension)

**Comment by whannah1** [comment]
@jayeshkrishna that's correct. I talked to Gautum about this and he wasn't too surprised because the feature that disables ELM history files is used while also disabling restarts. So this is a "bug" that no one has noticed until now because no one wanted to debug without history but with restarts.

**Comment by ndkeen** [comment]
The original reason to create this issue was that I had trouble CONTINUEing a run from restart, but it's quite likely that the restarts were not being fully written out and I tried to CONTINUE anyway.  

We have been debugging issues writing the restarts and have documented in other places. Using one of the solutions found (in this case 3 libfabric env vars), I can write restarts without a problem.
https://github.com/E3SM-Project/E3SM/pull/5275

I just ran a 6-month ne30 case with restarts every 2 months.  I then CONTINUEed twice, resulting in 1.5 years of simulation and did not see a problem.

So while we don't have a solution to the hanging restart writes in the repo, I wanted to close this particular issue as I think using restarts seems to be working fine assuming a complete set of restart files are written.
